Implement GH-7922: print_r should not produce a fatal error

mvorisek opened the pull request:
fixes https://github.com/php/php-src/issues/7922, impl. based on https://github.com/php/php-src/pull/7964

Fataling if `::__debugInfo()` doesn't return an array appears to be overly restrictive; we can handle that like returning null, namely by throwing an `Error` exception and returning an empty array instead.

PS: if null is returned, we don't throw an exception like before. Exception is only thrown if not array|null.

mvorisek commented on Zend/zend_object_handlers.c:
```diff
@@ -164,27 +164,30 @@ ZEND_API HashTable *zend_std_get_debug_info(zend_object *object, int *is_temp) /
+				return Z_ARRVAL(retval);
+			}
+		default:
+			zend_throw_error(NULL, ZEND_DEBUGINFO_FUNC_NAME "() must return an array");
```
this code should not be reached in case of an exception has been thrown

also, the `print_r`/`var_dump` statement should not print anything when an exception has been raised

arnaud-lb commented:
Thank you @mvorisek 

There are a few failing tests relating to __debugInfo

mvorisek commented:
@arnaud-lb credit should go to @cmb69

I need help with this PR, I do not know all the php-src internals

iluuu1994 reviewed (commented):
Two things are unfortunate about the implementation:

* When not passing the `$return = true` argument the exception occurs when all the previous values (and even the object header `<<class>> Object`) have already been dumped.
* The `get_debug_info` so far is not built for throwing exceptions, all places that use it (including `zend_get_properties_for`) will now need to check for `EG(exception)`

I'm not sure how we should address these. As for 1, buffering is not feasible for large dumped values. As for 2, we could document that extensions need to check for `EG(exception)` or adjust the handler to return `zend_result`. But I'm not sure this is worth it.

mvorisek commented on Zend/tests/gh7922.phpt:
```diff
@@ -0,0 +1,99 @@
+#2 {main}
+
+Exception occuring later on:
+Array
```
I expect no output from `print_r`/`var_dump` in case an exception occured during the printing - the whole output should be therefore buffered and printed only once it is complete without any exception.

mvorisek commented:
@iluuu1994 I have rebased the commits and fixed one failing ffi test. One test is remaining - can you please look at it?

I think there should be no partial output:
- object name /wo properties is misleading
- apps cannot change HTTP headers once output is send to the client (if output is not buffered), so the possible print_r/var_dump exception cannot be handled and sent to the client /w 500 HTTP code

iluuu1994 commented:
Well, have you solved the issues in https://github.com/php/php-src/pull/8865#pullrequestreview-1048671310? Because I don't know how to do that.

iluuu1994 commented:
Well, have you solved the issues in https://github.com/php/php-src/pull/8865#pullrequestreview-1048671310? Because I don't know how to do that.

> IMHO it should be fine to buffer the whole result, large/almost infine print_r output is usually a recursive references problem

That still means many changes across the file and in extensions. Extensions will break, as otherwise the will print their output to the output buffer while print_r will print to the tmp buffer.

mvorisek commented:
> That still means many changes across the file and in extensions. Extensions will break, as otherwise the will print their output to the output buffer while print_r will print to the tmp buffer.

- `print_r` has 2nd param to buffer/return the output, so `print_r` should be easily fixable (process like if 2nd param was true and echo at the end)
- `var_dump` - by definition, when we need to output nothing if any exception is emit (especially when processing an array), we must buffer the output, how much is it a problem for extensions when merged into `master`?

mvorisek commented:
I would prefer no output but if the extra memory needed is a problem, let's go output `*EXCEPTION*` the same as `*RECURSION*` for recursive data.

https://3v4l.org/DWqcW

also this is how `echo` with multiple arguments behave

iluuu1994 commented:
This has one stale. The comments from https://github.com/php/php-src/pull/8865#pullrequestreview-1048671310 stil need some solution, but I don't have one that is satisfactory. Feel free to open a new PR if you find a solution.

mvorisek commented:
> As for 1, buffering is not feasible for large dumped values.

The original issue is for `print_r` and it seems it is buffered for whole output already - https://github.com/php/php-src/blame/331ac35f5802145648fefa6312aa54c7306b93d6/Zend/zend.c#L628.
